### ISSUE TITLE
Incorrect Performance comparison

### ISSUE THREAD

**Issue by mdevils** [fb55/entities#377]
Hello guys, someone linked me to your performance comparison (https://github.com/fb55/entities#performance) and I was confused. Both by the results and by the benchmark itself.

I've created a different benchmark to collect more statistically significant data: https://github.com/mdevils/html-entities#performance

It would be nice if you made your performance comparison more correct.

**Comment by fb55** [comment]
Hey @mdevils, thanks for reaching out, and exciting to see another benchmark for this category of modules!

I like the inclusion of a spread here, but am failing to see where statistical significance comes to play. Which `p` value are you using?

Also congrats on the 2.0 release! I will update my benchmark to use the latest version of your module, very curious how things have changed.

**Comment by mdevils** [comment]
Thanks!

There are a couple of things to mention:

1. The benchmark I've provided uses `benchmark.js` which produces way more stable output.
2. Unless I'm missing something the benchmark used in `entities` uses extremely short input which generates a big deviation. The longer input you use, the more accurate results you get.

**Comment by fb55** [comment]
The benchmark in this repo is using 10,000,000 iterations for each library. The law of big numbers means that this will converge to a very stable result. It looks like the `html-entities` benchmark is using less than 100 iterations for each library, which means that (1) one will see a large spread and a lot of divergence between runs and (2) is is likely testing interpreter performance, not JIT performance. Especially (2) is counter to what one will actually see when using the library in the real world.

As to using smaller inputs: Having a smaller input means we measure initialization costs a lot more. It would be interesting to see what the average string length in an HTML document is. I would be surprised if this is much larger than the 31 characters of this repo's benchmark.

The `html-entities` benchmark is also doing a bit of an apples to oranges comparison, as it compares different classes of encoding as being in the same category.

**Comment by mdevils** [comment]
> The html-entities benchmark is also doing a bit of an apples to oranges comparison, as it compares different classes of encoding as being in the same category.

Yep, but I'm not making conclusions here, I just put everything together and show the performance.

**Comment by mdevils** [comment]
What is important for me is that what I'm getting in my results are not even close to to the results published on your README file.

**Comment by mdevils** [comment]
Not exactly. It sets a timer a sees how many executions fits in that time span. Here are the real run numbers for the first test for instance:

```js
{ 'html-entities.encode - html5, specialChars': 6341392,
  'html-entities.encode - html5, nonAsciiPrintable': 2212915,
  'html-entities.encode - html5, nonAscii': 2297872,
  '(old) Html4Entities.encode': 792195,
  '(old) Html5Entities.encode': 735717,
  'entities.encodeHTML5': 689013,
  'he.encode': 621028 }
```

**Comment by mdevils** [comment]
I think I see the problem in your benchmark now.

```
        (str: string) => htmlEntities.AllHtmlEntities.encode(str),
```

Static `encode` method was left for compatibility and extermely unperformant. In order to do a proper benchmark you need to create an instance of `AllHtmlEntities` and call `encode` on that instance.

**Comment by mdevils** [comment]
I've broken down my benchmark into comparable categories:

```
HTML5

    Encode test

        #1: html-entities.encode - html5, nonAscii x 362,978 ops/sec ±1.10% (93 runs sampled)
        #2: (old) Html5Entities.encodeNonASCII x 342,831 ops/sec ±0.94% (91 runs sampled)
        #3: html-entities.encode - html5, nonAsciiPrintable x 343,554 ops/sec ±1.35% (93 runs sampled)
        #4: entities.encodeHTML5 x 106,418 ops/sec ±0.41% (91 runs sampled)
        #5: he.encode x 98,762 ops/sec ±0.93% (92 runs sampled)

    Decode test

        #1: html-entities.decode - html5, strict x 291,466 ops/sec ±3.09% (86 runs sampled)
        #2: html-entities.decode - html5, body x 293,192 ops/sec ±5.33% (91 runs sampled)
        #3: html-entities.decode - html5, attribute x 272,764 ops/sec ±2.30% (91 runs sampled)
        #4: entities.decodeHTML5 x 250,868 ops/sec ±0.59% (93 runs sampled)
        #5: (old) Html5Entities.decode x 199,830 ops/sec ±5.65% (88 runs sampled)
        #6: he.decode x 176,291 ops/sec ±4.48% (88 runs sampled)

HTML4

    Encode test

        #1: html-entities.encode - html4, nonAscii x 364,044 ops/sec ±1.70% (92 runs sampled)
        #2: (old) Html4Entities.encodeNonASCII x 346,541 ops/sec ±2.07% (95 runs sampled)
        #3: html-entities.encode - html4, nonAsciiPrintable x 328,646 ops/sec ±2.61% (88 runs sampled)
        #4: entities.encodeHTML4 x 105,841 ops/sec ±0.53% (91 runs sampled)

    Decode test

        #1: html-entities.decode - html4, body x 327,197 ops/sec ±0.76% (95 runs sampled)
        #2: html-entities.decode - html4, attribute x 317,243 ops/sec ±2.10% (90 runs sampled)
        #3: html-entities.decode - html4, strict x 312,160 ops/sec ±2.94% (83 runs sampled)
        #4: entities.decodeHTML4 x 288,001 ops/sec ±0.85% (93 runs sampled)
        #5: (old) Html4Entities.decode x 245,895 ops/sec ±2.71% (90 runs sampled)

XML

    Encode test

        #1: (old) XmlEntities.encodeNonASCII x 1,174,893 ops/sec ±0.88% (91 runs sampled)
        #2: html-entities.encode - xml, nonAscii x 427,148 ops/sec ±1.40% (94 runs sampled)
        #3: html-entities.encode - xml, nonAsciiPrintable x 397,820 ops/sec ±2.77% (92 runs sampled)
        #4: entities.encodeXML x 283,177 ops/sec ±3.57% (93 runs sampled)

    Decode test

        #1: html-entities.decode - xml, strict x 407,364 ops/sec ±0.55% (91 runs sampled)
        #2: (old) XmlEntities.decode x 407,054 ops/sec ±0.50% (93 runs sampled)
        #3: entities.decodeXML x 405,984 ops/sec ±0.78% (94 runs sampled)
        #4: html-entities.decode - xml, body x 402,167 ops/sec ±0.67% (95 runs sampled)
        #5: html-entities.decode - xml, attribute x 387,630 ops/sec ±2.51% (94 runs sampled)
        #6: entities.decodeXMLStrict x 390,023 ops/sec ±7.37% (86 runs sampled)

Escaping

    Escape test

        #1: (old) XmlEntities.encode x 1,188,398 ops/sec ±0.83% (94 runs sampled)
        #2: he.escape x 1,163,620 ops/sec ±2.53% (88 runs sampled)
        #3: html-entities.encode - xml, specialChars x 1,087,350 ops/sec ±4.06% (91 runs sampled)
```

**Comment by mdevils** [comment]
@fb55 any update here?

**Comment by fb55** [comment]
Just started digging into this, to understand where the differences are coming from. https://github.com/mdevils/html-entities/issues/50 is a bug in `html-entities` that will lead to a speed-boost.

I have a preference to update the benchmark here once it actually is an apples-to-apples comparison. I know that `entities` currently doesn't provide the necessary APIs to accomplish some of the comparisons and started adding them with the latest release.

**Comment by mdevils** [comment]
@fb55, thanks for the discovery. The bug is fixed in `2.0.5`.

**Comment by mdevils** [comment]
Benchmark results so far:

```
HTML5

    Encode test

        #1: html-entities.encode - html5, nonAscii x 420,957 ops/sec ±0.32% (93 runs sampled)
        #2: html-entities.encode - html5, nonAsciiPrintable x 411,321 ops/sec ±0.38% (94 runs sampled)
        #3: (old) Html5Entities.encodeNonASCII x 354,640 ops/sec ±0.40% (90 runs sampled)
        #4: entities.encodeHTML5 x 121,672 ops/sec ±0.40% (95 runs sampled)
        #5: he.encode x 111,524 ops/sec ±0.35% (95 runs sampled)

    Decode test

        #1: html-entities.decode - html5, attribute x 331,117 ops/sec ±0.25% (95 runs sampled)
        #2: html-entities.decode - html5, strict x 332,711 ops/sec ±1.21% (94 runs sampled)
        #3: html-entities.decode - html5, body x 328,533 ops/sec ±0.45% (94 runs sampled)
        #4: entities.decodeHTML5 x 270,985 ops/sec ±1.05% (96 runs sampled)
        #5: (old) Html5Entities.decode x 223,444 ops/sec ±0.59% (95 runs sampled)
        #6: he.decode x 182,908 ops/sec ±0.51% (94 runs sampled)

HTML4

    Encode test

        #1: html-entities.encode - html4, nonAscii x 399,177 ops/sec ±0.28% (94 runs sampled)
        #2: html-entities.encode - html4, nonAsciiPrintable x 378,698 ops/sec ±0.98% (91 runs sampled)
        #3: (old) Html4Entities.encodeNonASCII x 362,195 ops/sec ±0.51% (90 runs sampled)
        #4: entities.encodeHTML4 x 117,798 ops/sec ±0.21% (95 runs sampled)

    Decode test

        #1: html-entities.decode - html4, body x 358,360 ops/sec ±0.23% (96 runs sampled)
        #2: html-entities.decode - html4, strict x 357,440 ops/sec ±1.12% (91 runs sampled)
        #3: html-entities.decode - html4, attribute x 354,844 ops/sec ±0.54% (96 runs sampled)
        #4: entities.decodeHTML4 x 303,150 ops/sec ±0.37% (94 runs sampled)
        #5: (old) Html4Entities.decode x 247,929 ops/sec ±1.34% (94 runs sampled)

XML

    Encode test

        #1: (old) XmlEntities.encodeNonASCII x 1,119,454 ops/sec ±1.04% (96 runs sampled)
        #2: html-entities.encode - xml, nonAscii x 454,423 ops/sec ±0.70% (95 runs sampled)
        #3: html-entities.encode - xml, nonAsciiPrintable x 440,705 ops/sec ±0.28% (95 runs sampled)
        #4: entities.encodeXML x 338,496 ops/sec ±0.43% (94 runs sampled)

    Decode test

        #1: entities.decodeXMLStrict x 425,734 ops/sec ±0.27% (94 runs sampled)
        #2: html-entities.decode - xml, body x 420,139 ops/sec ±0.29% (95 runs sampled)
        #3: entities.decodeXML x 416,914 ops/sec ±1.30% (93 runs sampled)
        #4: html-entities.decode - xml, strict x 416,223 ops/sec ±1.23% (93 runs sampled)
        #5: (old) XmlEntities.decode x 417,082 ops/sec ±1.52% (94 runs sampled)
        #6: html-entities.decode - xml, attribute x 403,015 ops/sec ±0.41% (93 runs sampled)

Escaping

    Escape test

        #1: he.escape x 1,119,507 ops/sec ±0.49% (94 runs sampled)
        #2: (old) XmlEntities.encode x 1,115,930 ops/sec ±1.02% (93 runs sampled)
        #3: html-entities.encode - xml, specialChars x 1,068,475 ops/sec ±0.47% (96 runs sampled)
```

**Comment by mdevils** [comment]
Note that in the last benchmark I've also added initialization / loading speed comparison.

**Comment by fb55** [comment]
Hi @mdevils, sorry for the pause — I wanted to take some time to understand why the libraries are performing so differently before deciding what to do here. Given how long this took, I should have removed the benchmark from the README in the meantime.

After looking into this some more, I have to say that I was quite surprised to see how much of a change in performance your specialized `String.prototype.replace` variant brings. In this repo's basic benchmarks, it leads to ~50% gains. Awesome find, I would not have thought about hand-rolling this.

Would you mind if I publish the `replace` implementation as a separate package, that I can use for `entities`? I am sure there are other packages that will benefit from it. I will make sure you get the credit you deserve.

**Comment by mdevils** [comment]
@fb55, it's OpenSource, do whatever you want. For quite a long time already I'm just asking to finally stop this misinformation and delete incorrect benchmarks.

**Comment by fb55** [comment]
@mdevils Done in #560. Thanks for kicking my butt, this was a good reason to dive deep into string algos 😄

**Comment by fb55** [comment]
@mdevils Released my changes as `3.0.0`, and gave you a shout-out in the README and the release notes: https://github.com/fb55/entities/releases/tag/v3.0.0

Curious to see how long it'll take for you to update your benchmark 😉
